observer/probers/crl/crl_conf.go

```diff
@@ -0,0 +1,132 @@
+package probers
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/letsencrypt/boulder/observer/probers"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	nextUpdateName = "obs_crl_next_update"
+	thisUpdateName = "obs_crl_this_update"
+	certCountName  = "obs_crl_revoked_cert_count"
+)
+
+// CRLConf is exported to receive YAML configuration
+type CRLConf struct {
+	URL string `yaml:"url"`
+}
+
+// Kind returns a name that uniquely identifies the `Kind` of `Configurer`.
+func (c CRLConf) Kind() string {
+	return "CRL"
+}
+
+// UnmarshalSettings constructs a CRLConf object from YAML as bytes.
+func (c CRLConf) UnmarshalSettings(settings []byte) (probers.Configurer, error) {
+	var conf CRLConf
+	err := yaml.Unmarshal(settings, &conf)
+	if err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+func (c CRLConf) validateURL() error {
+	url, err := url.Parse(c.URL)
+	if err != nil {
+		return fmt.Errorf(
+			"invalid 'url', got: %q, expected a valid url", c.URL)
+	}
+	if url.Scheme == "" {
+		return fmt.Errorf(
+			"invalid 'url', got: %q, missing scheme", c.URL)
+	}
+	return nil
+}
+
+// MakeProber constructs a `CRLProbe` object from the contents of the
+// bound `CRLConf` object. If the `CRLConf` cannot be validated, an
+// error appropriate for end-user consumption is returned instead.
+func (c CRLConf) MakeProber(collectors map[string]prometheus.Collector) (probers.Prober, error) { // validate `url` err := c.validateURL()
+	// validate `url`
+	err := c.validateURL()
+	if err != nil {
+		return nil, err
+	}
+	probe := CRLProbe{c.URL, nil, nil, nil}
+
+	// validate the prometheus collectors that were passed in
+	var ok bool
+	for name, coll := range collectors {
+		switch name {
+		case nextUpdateName:
+			probe.cNextUpdate, ok = coll.(*prometheus.GaugeVec)
+			if !ok {
+				return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", name, coll)
+			}
+		case thisUpdateName:
+			probe.cThisUpdate, ok = coll.(*prometheus.GaugeVec)
+			if !ok {
+				return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", name, coll)
+			}
+		case certCountName:
+			probe.cCertCount, ok = coll.(*prometheus.GaugeVec)
+			if !ok {
+				return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", name, coll)
+			}
+		default:
+			return nil, fmt.Errorf("crl prober received unexpected collector %q", name)
+		}
+	}
+	if probe.cNextUpdate == nil {
+		return nil, fmt.Errorf("crl prober did not receive collector %q", nextUpdateName)
+	}
+	if probe.cThisUpdate == nil {
+		return nil, fmt.Errorf("crl prober did not receive collector %q", thisUpdateName)
+	}
+	if probe.cCertCount == nil {
+		return nil, fmt.Errorf("crl prober did not receive collector %q", certCountName)
+	}
```

Comment:
Using a loop here feels like a good idea, because it gives us a chance for code re-use. But then breaking things out with the switch/case removes all chance for code reuse anyway, so the loop isn't actually buying us anything. I think you can get the same result, in even fewer lines of easier-to-read code, by dropping the loop and the switch/case. This also lets us create the `probe` return object at the last second, so it isn't sitting around begging for a buggy refactoring to cause a partially-initialized object to be returned early:

```suggestion

	coll, ok := collectors[nextUpdateName]
	if !ok {
		return nil, fmt.Errorf("crl prober did not receive collector %q", nextUpdateName)
	}
	nextUpdateColl, ok := coll.(*prometheus.GaugeVec)
	if !ok {
		return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", nextUpdateName, coll)
	}

	coll, ok = collectors[nextUpdateName]
	if !ok {
		return nil, fmt.Errorf("crl prober did not receive collector %q", thisUpdateName)
	}
	thisUpdateColl, ok := coll.(*prometheus.GaugeVec)
	if !ok {
		return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", thisUpdateName, coll)
	}

	coll, ok = collectors[certCountName]
	if !ok {
		return nil, fmt.Errorf("crl prober did not receive collector %q", certCountName)
	}
	certCountColl, ok := coll.(*prometheus.GaugeVec)
	if !ok {
		return nil, fmt.Errorf("crl prober received collector %q of wrong type, got: %T, expected *prometheus.GaugeVec", certCountName, coll)
	}

	probe := CRLProbe{c.URL, nextUpdateColl, thisUpdateColl, certCountColl}
```

Reply:
I think this makes sense, but I'm curious about the case in which `MakeProber()` might receive some number of collectors that it wasn't expecting (i.e. the `default` case in the switch statement). Do we think this is important to check for, and if so, how would we best check for that condition outside of the loop/switch structure?

I'll note that this question originally came up in an offline conversation w/ @beautifulentropy that led to the loop & switch method that you see here. My original, perhaps naive, suggestion was to do the checks much like you've suggested here, and for the aforementioned corner case, do a check of `len(collectors)` to make sure it's not greater than expected. Though it was pointed out that the number would be hardcoded and need to change if future updates to the code added/removed some collector(s). It also wouldn't necessarily give us a nice error message that includes the name of the  extraneous collector(s).

Maybe this is alright though. Or maybe we just don't find it super important to check for this condition in the first place. Thoughts?

Reply:
Yep, it's a good question! I think my answer is based on the following question: What's the harm in being passed a extra collector?

_Not_ being passed one of these collectors is a very bad thing, because it means a) a metric we expect to be able to observe won't be updated; and more importantly b) we'll get a nil pointer exception and panic when we try.

But being passed an extra one? Sure, it means someone wrote a bug somewhere, but fundamentally the process will continue to behave correctly. I don't think it's worth testing for that at runtime.

It may be worth adding a test which does the basic setup and then asserts that the map which gets passed to MakeProber has length 3. But I don't think that performing this check at runtime provides enough value to be worth the extra nesting and complication of the surrounding code.

Reply:
Yeah I think that's a pretty reasonable way to look at it. And arguably, there's no need to even make that extra test you mention. I think what's most important to us is not that `MakeProber()` is passed a map of collectors of a very specific size, but more that it doesn't error out when passed the map returned from `Instrument()` (which *will* necessarly imply some notion of "there must be at least X collectors in this map", but, as you said, there's really no harm if there are *more than* X for some reason). The test for `MakeProber()` already checks that the output of `Instrument()` doesn't cause `MakeProber()` to error out, so I think that's sufficient.